modules/3d/src/rgbd/volume_impl.cpp

```diff
@@ -213,6 +513,216 @@ int TsdfVolume::getVisibleBlocks() const { return 1; }
 size_t TsdfVolume::getTotalVolumeUnits() const { return 1; }
 
 
+template <typename T, size_t K>
+bool vec_cmp(const Vec<T, K>& a, const Vec<T, K>& b)
+{
+    return std::lexicographical_compare(&a[0], &a[0] + K, &b[0], &b[0] + K);
+}
+
+template <typename T, size_t K>
+bool unordered_vec_cmp(const Vec<T, K>& a, const Vec<T, K>& b)
+{
+    Vec<T, K> ta = a, tb = b;
+    std::sort(&ta[0], &ta[0] + K);
+    std::sort(&tb[0], &tb[0] + K);
+    return std::lexicographical_compare(&ta[0], &ta[0] + K, &tb[0], &tb[0] + K);
+}
+
+struct MarchCubesInvoker : ParallelLoopBody
+{
+    MarchCubesInvoker(const TsdfVolume& _volume,
+                      const VolumeSettings& _settings,
+                      std::vector<Vec3f>& _meshPoints,
+                      std::vector<Vec3i>& _meshFaces) :
+            volume(_volume),
+            settings(_settings),
+            mcNeighbourPts{
+                    Point3f(0.f, 0.f, 0.f),
+                    Point3f(0.f, 0.f, 1.f),
+                    Point3f(0.f, 1.f, 1.f),
+                    Point3f(0.f, 1.f, 0.f),
+                    Point3f(1.f, 0.f, 0.f),
+                    Point3f(1.f, 0.f, 1.f),
+                    Point3f(1.f, 1.f, 1.f),
+                    Point3f(1.f, 1.f, 0.f)
+            },
+            mcNeighbourCoords([&_settings]{
+                const Vec4i volDims;
+                _settings.getVolumeDimensions(volDims);
+                return Vec8i(
+                         volDims.dot(Vec4i(0, 0, 0)),
+                         volDims.dot(Vec4i(0, 0, 1)),
+                         volDims.dot(Vec4i(0, 1, 1)),
+                         volDims.dot(Vec4i(0, 1, 0)),
+                         volDims.dot(Vec4i(1, 0, 0)),
+                         volDims.dot(Vec4i(1, 0, 1)),
+                         volDims.dot(Vec4i(1, 1, 1)),
+                         volDims.dot(Vec4i(1, 1, 0))
+                        );
+            }()),
+            volData(_volume.volume.ptr<TsdfVoxel>()),
+            meshPoints(_meshPoints),
+            meshFaces(_meshFaces)
+            {}
+
+    Point3f interpolate(Point3f p1, Point3f p2, float v1, float v2) const
+    {
+        float dV = 0.5f;
+        if (abs(v1 - v2) > 0.0001f)
+            dV = v1 / (v1 - v2);
+
+        Point3f p = p1 + dV * (p2 - p1);
+        return p;
+    }
+
+    virtual void operator()(const Range &range) const override
+    {
+        Vec4i volDims;
+        settings.getVolumeDimensions(volDims);
+
+        Vec3i resolution;
+        settings.getVolumeResolution(resolution);
+        const Point3i volResolution = Point3i(resolution);
+
+        Matx44f _pose;
+        settings.getVolumePose(_pose);
+        const Affine3f pose = Affine3f(_pose);
+
+        float voxelSize = settings.getVoxelSize();
+
+        std::map<Vec3f, size_t, decltype(&vec_cmp<float, 3>)> point_id(vec_cmp<float, 3>);
```

Comment:
BTW looks like there's a float-float comparison happens which (usually) works well but (ideally) should be done with epsilons. I propose to use integer keys here if possible.